File: sdk-python/src/daytona/common/warm_pool.py
from __future__ import annotations

from daytona_api_client import WarmPool as WarmPoolDto
from daytona_api_client_async import WarmPool as AsyncWarmPoolDto


class WarmPool(WarmPoolDto):
    """Represents a Daytona Warm Pool which keeps ready-to-use Sandboxes for a snapshot.

    ``current_size`` versus ``pool`` is the pool's status: ``current_size`` is the number
    of ready warm sandboxes, ``pool`` is the desired number. ``error_reason`` is set when
    the pool cannot be filled.

    Attributes:
        id (str): Unique identifier for the Warm Pool.
        organization_id (str): Organization ID that owns the Warm Pool.
        snapshot (str): Snapshot the pool keeps warm sandboxes for.
        target (str): Target region of the pool.
        pool (int): Desired number of warm sandboxes.
        current_size (int): Current number of ready warm sandboxes in the pool.
        cpu (int): CPU cores per sandbox.
        mem (int): Memory per sandbox in GiB.
        disk (int): Disk per sandbox in GiB.
        os_user (str): OS user of the warm sandboxes.
        env (dict[str, str]): Environment variables of the warm sandboxes.
        error_reason (str | None): Reason the pool cannot be filled, if any.
        created_at (str): Date and time when the Warm Pool was created.
        updated_at (str): Date and time when the Warm Pool was last updated.
    """

    @classmethod
    def from_dto(cls, dto: WarmPoolDto | AsyncWarmPoolDto) -> "WarmPool":
        return cls.model_validate(dto.model_dump())

File: sdk-ruby/lib/daytona/warm_pool.rb
# frozen_string_literal: true

module Daytona
  # A warm pool of ready-to-use Sandboxes for a snapshot. `current_size` versus
  # `pool` is the pool's status: `current_size` is the number of ready warm
  # sandboxes, `pool` is the desired number. `error_reason` is set when the
  # pool cannot be filled.
  class WarmPool
    # @return [String]
    attr_reader :id

    # @return [String]
    attr_reader :organization_id

    # @return [String]
    attr_reader :snapshot

    # @return [String]
    attr_reader :target

    # @return [Integer]
    attr_reader :pool

    # @return [Integer]
    attr_reader :current_size

    # @return [Integer]
    attr_reader :cpu

    # @return [Integer]
    attr_reader :mem

    # @return [Integer]
    attr_reader :disk

    # @return [String]
    attr_reader :os_user

    # @return [Hash{String => String}]
    attr_reader :env

    # @return [String, nil]
    attr_reader :error_reason

    # @return [String]
    attr_reader :created_at

    # @return [String]
    attr_reader :updated_at

    # Initialize warm pool from DTO
    #
    # @param warm_pool_dto [DaytonaApiClient::WarmPool]
    def initialize(warm_pool_dto)
      @id = warm_pool_dto.id
      @organization_id = warm_pool_dto.organization_id
      @snapshot = warm_pool_dto.snapshot
      @target = warm_pool_dto.target
      # The generated client deserializes OpenAPI numbers as Float; these are all integral.
      @pool = warm_pool_dto.pool.to_i
      @current_size = warm_pool_dto.current_size.to_i
      @cpu = warm_pool_dto.cpu.to_i
      @mem = warm_pool_dto.mem.to_i
      @disk = warm_pool_dto.disk.to_i
      @os_user = warm_pool_dto.os_user
      @env = warm_pool_dto.env
      @error_reason = warm_pool_dto.error_reason
      @created_at = warm_pool_dto.created_at
      @updated_at = warm_pool_dto.updated_at
    end
  end
end
